### ISSUE TITLE
Excel triggers calculation of non-volatile custom functions with certain actions like expand group or insert row

### ISSUE THREAD

**Issue by 4tti** [OfficeDev/office-js#3332]
# Provide required information needed to triage your issue
Certain actions in excel trigger recalculation of non-volatile functions although they should be treated as non-dirty and they don't need any recalculation.
Identified actions so far:
- expanding/collapsing grouped cells
- inserting row/column
- deleting row/column
- inserting/deleting cell

This is crucial for calculation performance. Imagine case when report designer hides some technical stuff (referenced by formulas) using grouping and just expanding it will trigger recalculation of the sheet. It gets even worse with complex reports when there are cross sheet references.
See video below.

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: PC desktop
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: 
![image](https://user-images.githubusercontent.com/19533927/235940755-af6604ae-0ea4-445a-bffc-b55e74b4d98a.png)
* Operating System: Windows
* Browser (if using Office on the web): ______
 
## Expected behavior

Non-volatile custom functions should be recalculated only once needed or marked as dirty.

## Current behavior

Non-volatile custom functions are always recalculated unnecessarily

## Steps to reproduce

1. See video below.
2. One formula is custom function (non-volatile)
3. The other is sum native function (non-volatile AFAIK).
4. See the difference in behavior

## Link to live example(s)

1. https://user-images.githubusercontent.com/19533927/235941288-8c51e84a-1f62-4b19-8550-9af2308fa9f8.mp4

**Comment by gmichaud** [comment]
We noticed the same and hope this can be avoided!

**Comment by rundongmsft** [comment]
Hi @4tti, this is a by-design behavior. Custom function is recognized as array formula by Excel even though it returns one value. In fact, the native array formula has the same behavior. You can try the following step:
1. Insert a native array formula at G5, for example: “=COLUMN(R1)”.
2. Insert a column at H.
3. You will see that the formula at G5 is recalculated.

**Comment by 4tti** [comment]
@rundongmsft  sorry but your usage is not correct because simply inserting the column will CHANGE the value to be calculated and so it is really dirty.

However, when you insert in G5 "=COLUMN(H5)" and then you insert column in Z -> no calculation happens.
So please check again.

**Comment by rundongmsft** [comment]
Hi @4tti, `when you insert in G5 "=COLUMN(H5)" and then you insert column in Z` the calculation happens. Since it's calculated quickly and the result doesn't change, it's hard to realize that it's recalculated. And CF has a stage of "#busy", so it is easy to observe that there is a recalculation.
In fact, [Deleting or inserting a row or column will trigger recalculation in Excel](https://learn.microsoft.com/en-us/office/client-developer/excel/excel-recalculation) when the Calculation Options is "Automatic" (by default). Even if the recalculated formulas or their references are not dirty. This situation does not only happen in array formulas, but also in others (e.g. SUM). Sorry I only mentioned array formula earlier.

**Comment by 4tti** [comment]
Ok, you are right it is happening for at least COM/VBA user defined functions, I cannot judge if it affects even native functions.
However, here is one crucial impact you should consider: 
Each calculation of Custom Functions usually means web request and so effectively has these consequences:
1. because of the optimizations we collect/bulk the formulas -> we don't want to flood service with thousands of requests 
2. we send the request and process it on the service
3. we receive "updated" value.

So effectively inserting one line can - with automatic calculation - cause 10s delay for the user.

How can I explain that to the customers who complain about it? 
They argue that with native or COM based UDFs it is not happening (which is "true" beause it is almost invisible because of the speed). 

**Could we figure out some good solution for this? Because it is really painful for the customers.**
I proposed to use manual calculation but it has some side-effects + it needs users to have some 'knowledge' about it.

We can - as last resort - introduce switch to manual calculation in our add-in, but it is not ideal and can cause some unwanted consequences as well (+ it is behavior change).

**Comment by 4tti** [comment]
@rundongmsft  can we get some solution please? This is really painful experience for multiple customers.

**Comment by jiju-MS** [comment]
Hey, @4tti , as @rundongmsft mentioned, this behavior is controlled by the excel calculation system so there is nothing custom function can do about it. The only way to avoid it is to use manual calculation. 

One possible way we catch up may reduce the pain of manual calculation is that you can use data change event and formular change event to monitor the change on worksheet, then you can call "calculate" api in the event callback to help user do the calculation instead of let them do it by themselves. 

when user inserting row/column or deleting row/column, the data change address will be like 1:1, which is easy to recognize and filter, in that way you may avoid the issue. 

May this be of some help to you.

**Comment by 4tti** [comment]
Thanks a lot @jiju-MS, but no - we won't override excel behavior that much + how should we prevent user of changing it to automatic? Should we even prevent it?

I understand it is controlled by excel calculation engine... but I am asking you to consider this scenario (custom functions sending web requests) as one of the main reasons why Web Add-ins are perceived as extremely inefficient compared to COM add-ins.
When you work with small reports (tens/hundreds of CFs) and you have good network - it is OK.
But when you start having (tens of) thousands of CFs or you have a bit slow network -> it is PAINFUL. 

To give whole view - the other reason is nesting - when you e.g. chain multiple CFs the calculation prolongs significantly because of the web requests and calc dependency.

One idea how to improve it -> introduce some 'immutable' type (similar to volatile) -> which would not be recalculated with such events - only when MANUALLY marked as 'dirty'
Or eventually introduce range property which could drive this (i.e. SkipRecalculation or so).

This way we would be able to significantly boost the performance. Just an idea.

**Comment by rundongmsft** [comment]
Hi @4tti, thanks for your suggestion, it's a good idea to solve this issue. We have created a internal product backlog 8189589 to track this.

**Comment by asegal-singlepane** [comment]
Hello - we have the exact same issue with our custom functions which call a webservice/api.  We were about to implement batching to avoid overloading our webservice and timing out calls, but realize reading this that won't really solve our issue in that people often add/delete rows etc to reports with our functions and we can't trigger 1000s of calls every time someone does that.  Has anyone thought of any clever workarounds or has there been any evolution on the MS side?

**Comment by asegal-singlepane** [comment]
@4tti did you make any adjustments to your UDFs to improve this?

**Comment by 4tti** [comment]
@asegal-singlepane no adjustments after reporting this bug.
We have introduced batching with initial versions of our add-in + changed formulas removing volatility (As it was nightmare) and introduced 'custom calculation' - i.e. we have our own buttons to trigger calculation of our custom functions. Not nice and with this bug still ugly.

Customers are still complaining about this behavior... we don't want to hack excel, so we are still eagerly awaiting some improvements in the Calc Engine from MS.

**Comment by asegal-singlepane** [comment]
Thanks for the reply @4tti. We’re going to implement batching now and also looking at creating a local cache either in local storage or within the document with custom xml so that even if the sheet refreshes on a structure change (added row) the recalc is fast because it can pull from the cache and not make a network call. Thinking together with batching this will make the behavior manageable.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jpickles1981** [comment]
I have a similar problem and will be exploring batching and caching using indexedDB / local storage.  However, I think there is another solution which I have described [here](https://techcommunity.microsoft.com/idea/microsoft365developerplatform/retrieve-the-existing-contents-of-a-cell-that-invokes-a-function/4449737).

It involves a function being able to check the contents of the very cell where the function is written.  This helps when the data is stored in the cell as Excel.EntityCellType, because we can store both API request (input) and response (output) in the cell.  It's effectively like a cell having its very own cache inside the cell - which I imagine is far more efficient than checking a browser's local storage.

How about that for a concept: a cell having its own cache!

**Comment by wh1t3cAt1k** [comment]
We use the calculation cache, and that helps somewhat.

However, our customers are still irked by the fact that the #BUSY refresh indicator flashes for a bit whenever they expand a row.

In our legacy COM product, we could **synchronously** return a value from the cache upon recalculation, which never resulted in #BUSY (or #N/A in case of `ExcelDna`), so the fact of recalculation was opaque to the customer.

@jiju-MS @rundongmsft @Wenjun-Gong – I have two proposals that do not necessarily contradict each other:

1. A simple-to-implement idea with immediate UX improvements – an **opt-in** "grace period" for custom functions. For ISVs that implemented the cell-level cache (like we did), expose an API to set up a delay **before showing #BUSY** or a refresh indicator. It could be a configurable numeric value for up to 0.5 seconds, for example. Anything that returns from the cache will not trigger #BUSY and thus not irritate the customer.

2. A more complex idea – support CF result caching at the framework level. For example, the add-in JS registers a special function called `customFunctionCacheLookup(invocation, customFunctionName, customFunctionArgumentArray)`, and this function returns `{ found: true, result: any }`, the #BUSY state is not entered at all. If it doesn't, then the CF's "normal" implementation is called.

**Comment by wh1t3cAt1k** [comment]
I didn't read @4tti 's idea carefully enough, but introducing a global setting for  stricter recalculation behaviour for CFs would also work for us. 

Our reporting workflow depends on the user manually clicking Velixo's "Refresh" button – it marks the worksheet contents as dirty **OR** in some cases, the application's "full calculation" (which should also mark everything as dirty).

**Comment by gmichaud** [comment]
To illustrate how recalculation behavior influences user perception when transitioning from our COM/XLL-based add-in to the OfficeJS version, here’s a support ticket we received this week:

> Hi - We are rtying to fully switch. But we are consistently caught in refresh hell. I'm not sure why - maybe its becasue we run pretty complex workbooks or seomthing. But it almost becomes unusable with the amount of refreshes it does. Almost as if as soon as you move a cell the whole friggin sheet/workbook refreshes....

**Comment by AmandaHan123** [comment]
Hi @gmichaud , @wh1t3cAt1k and @jpickles1981 , this is tracked by Microsoft internal backlog #8189589. I conveyed your words to the feature owner. We will reply you here if there is any update.

**Comment by ayush-meghwani-28** [comment]
Hello @4tti, thanks for the detailed report — this was really helpful.

## What's happening

When you insert or delete a row, expand/collapse a group, or otherwise change the
structure of a sheet, Excel recalculates the sheet. This is by design: structural
edits can move cells around and change dependencies, so Excel re-evaluates to keep
everything consistent.

For most functions this is invisible — they just return the same value again. But a
**request/response custom function** (one that returns a `Promise` and calls the
network each time it runs) makes a fresh network request every time it re-runs. So a
block of N such cells becomes N network requests on each structural edit — that's the
slowdown you're seeing. Marking the function non-volatile does **not** exempt it from
a *structural* recalc, so unfortunately it doesn't help here.

## A way to avoid the repeated requests

Write the function as a **streaming custom function** instead. The difference:

- **Request/response** (what you have today): every time Excel recalculates the cell,
  it runs your function again — so it fetches from the network again.
- **Streaming**: your function fetches **once**, gives Excel that value, and Excel keeps
  showing it. Excel will **not** run the function again just because you inserted a row
  or expanded a group. It only re-runs when one of the function's **input values**
  actually changes.

So with a streaming function, inserting rows no longer triggers new network requests.

### Before — request/response (fetches again on every structural recalc)

```ts
/**
 * @customfunction
 * @param {number} id
 * @returns {Promise<number>}
 */
async function getPrice(id: number): Promise<number> { const res = await fetch(`https://api.example.com/price/${id}`); const data = await res.json(); return data.price; // runs again on every insert-row / group expand
}
```

After — streaming (fetches once, kept across structural edits)

```ts
/**
 * @customfunction
 * @param {number} id
 * @param {CustomFunctions.StreamingInvocation<number>} invocation
 */
function getPrice(id: number, invocation: CustomFunctions.StreamingInvocation<number>): void { let cancelled = false;

  const load = async () => {
    const res = await fetch(`https://api.example.com/price/${id}`);
    const data = await res.json();
    if (!cancelled) {
      invocation.setResult(data.price); // give Excel the value once; it keeps showing it
    } };

  load();

  // Optional: refresh on your own schedule instead of on Excel's recalcs // const timer = setInterval(load, 60_000);

  invocation.onCanceled = () => {
    cancelled = true;
    // clearInterval(timer); };
}
```

(The  invocation.setResult(...)  call is how a streaming function hands Excel a value.
You decide when it happens — once after fetching, or on a timer if you want live
updates.)

The metadata is generated for you from the JSDoc — the  invocation  parameter is what
marks it as a streaming function — so you don't need to hand-edit  functions.json .

Notes / trade-offs

• Your function now decides when it refreshes (once after fetching, or on a timer),
instead of refreshing on Excel's structural recalcs. For "fetch once and hold" data
this is usually exactly what you want.
• A full/hard recalc (Ctrl+Alt+F9), or a genuine
change to an input value, will still trigger a fresh fetch — that's expected.
• If you want live updates, use the timer pattern above to refresh on your own cadence,
independent of sheet edits.

Separately, we're also looking into improving Custom Functions performance for these
structural (insert/delete/group) operations in Excel's calc engine, so that regular
(non-streaming) custom functions perform better in this scenario too. Hope this helps!

**Comment by 4tti** [comment]
@ayush-meghwani-28 Thanks a lot for your message!

Can you elaborate more on the 'push', what do you exactly mean by that?

**Comment by ayush-meghwani-28** [comment]
Hello there, sorry for the confusion and By "push" I just meant: a normal function waits to be asked and re-runs (re-fetches) every time Excel recalculates. A streaming function works the other way around — it fetches once, hands Excel that value, and Excel just keeps showing it. It only re-fetches when one of the function's inputs actually changes — not when you insert a row or expand a group. 
I've simplified my earlier message as well to say it this way.

**Comment by gmichaud** [comment]
@ayush-meghwani-28 thank you very much for the message! We are facing the same problem described by @4tti and given the performance of office-js custom functions it is generating complaints from users accustomed to the level of performance of our XLL-based addin (we have a function results cache and return the last value synchronously, avoiding any RTD/async overhead). 

Rewriting everything as streaming functions isn't a risk we're prepared to take on our end. We evaluated that path before and ran into a number of limitations; some, like #3100, have since been fixed, but the migration cost and remaining unknowns are still too high for a production add-in of our size.

We'd very much like to hear more about the improvements you're planning for regular, non-streaming custom functions in this scenario. Happy to share benchmarks or a minimal repro from our side if that would help prioritize the work.

**Comment by ayush-meghwani-28** [comment]
Hello @gmichaud Thank you — this is really helpful context, and your XLL comparison is a fair benchmark. I completely understand where you're coming from: your users are used to a certain level of responsiveness, and office-js custom functions aren't matching it. That's valid feedback and I'd like to dig into your specific setup rather than hand you generic advice.

A quick, honest note on the streaming suggestion: on this original reported issue, the pain was fairly narrow — a request/response function re-fetching on every structural edit (insert row, expand group) — and for *that* pattern I thought streaming's argument-based
caching might help, since it avoids the repeated calls. But you're right that streaming isn't a fix for everything, and it may not address your case at all. Your scenario looks similar on the surface, but it could be a bit different underneath. So instead of assuming, I'd rather reproduce
what you're actually seeing.

To do that accurately, could you share a few details about your office-js add-in setup?

1. **A minimal repro.** The smallest workbook + one representative function that shows the slowness — even pseudo-code of what it does is fine. In particular: does the function compute purely from its arguments, call the network, or read from a client-side cache (like your XLL does)?

2. **What triggers the worst slowdown**, roughly in order: row insert/delete, group expand/collapse, editing an input cell, full recalc (Ctrl+Alt+F9), or file open?

3. **Scale.** Roughly how many custom-function cells are calculating at once in a typical workbook, and in your worst case? This is the single most useful number for me to reproduce what you're experiencing.

4. **What "slow" means concretely** for you — the per-cell latency before a value appears, he total recalc time, the visible "calculating"/loading state, or overall responsiveness?

5. **Input stability.** Once computed, are a function's inputs mostly static, or do they change often? And does the result depend only on the arguments, or also on external  state (a data feed, time, the cell's own address)?

With a minimal repro and your scale numbers I can stand up an equivalent workbook on my side, measure it, and give you a straight read on whether anything in the currently-supported surface can meaningfully narrow the gap for your setup — and where it can't, your details would make the strongest possible case on our side. Really appreciate you working through this with us.

On a side note - If you don't mind, we can create a separate issue for your scenario to track it separately or else if you want to continue the conversation here only - that's also completely fine.

**Comment by gmichaud** [comment]
@ayush-meghwani-28 I recorded this video showing the behaviour on a test workbook, built against Velixo. Both the streaming and non-streaming version of the function has the same behaviour. In this example, you can see Excel start recalculating and it takes about 15 seconds before it settles: https://www.loom.com/share/e8ccc1ace6794772986148d53f19c7a0

You can replicate the same thing on your own add-in by simply having a 1 second wait per cell and returning a dummy value -- we have cached these values in our add-in so we're just returning the cached value.

VIdeo: https://www.loom.com/share/e8ccc1ace6794772986148d53f19c7a0

[Streaming functions.xlsx](https://github.com/user-attachments/files/31930768/Streaming.functions.xlsx)
[Regular functions.xlsx](https://github.com/user-attachments/files/31930769/Regular.functions.xlsx)

Can this GitHub issue be reopend please?
